clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ModifyServerGroupLaunchTemplateAtomicOperation.java

```diff
@@ -16,306 +16,59 @@
-          blockDevices = blockDevicesForInstanceType;
-        }
-      }
-    }
+  private List priorOutputs = new ArrayList();
 
-    LaunchTemplateVersion newVersion =
-        regionScopedProvider
-            .getLaunchTemplateService()
-            .modifyLaunchTemplate(
-                currentVersion,
-                description.getBase64UserData(),
-                newLaunchTemplateData,
-                Optional.ofNullable(blockDevices));
-
-    launchTemplateSpec.setVersion(String.valueOf(newVersion.getVersionNumber()));
-    LaunchTemplateSpecification newSpec =
-        new LaunchTemplateSpecification()
-            .withLaunchTemplateId(newVersion.getLaunchTemplateId())
-            .withVersion(String.valueOf(newVersion.getVersionNumber()));
-
-    // point asg to new template version
-    autoScaling.updateAutoScalingGroup(
-        new UpdateAutoScalingGroupRequest()
-            .withAutoScalingGroupName(description.getAsgName())
-            .withLaunchTemplate(newSpec));
-
-    return null;
-  }
-
-  private boolean matchingBlockDevices(
-      List<LaunchTemplateBlockDeviceMapping> mappings,
-      List<AmazonBlockDevice> blockDevicesForInstanceType) {
-    for (LaunchTemplateBlockDeviceMapping mapping : mappings) {
-      if (blockDevicesForInstanceType.stream()
-          .anyMatch(deviceForType -> !matchesDevice(deviceForType, mapping))) {
-        return false;
-      }
-    }
-
-    return true;
-  }
-
-  private boolean matchesDevice(
-      AmazonBlockDevice deviceForType, LaunchTemplateBlockDeviceMapping mapping) {
-    BlockDevice device1 =
-        new BlockDevice()
-            .withDeviceName(deviceForType.getDeviceName())
-            .withVirtualName(deviceForType.getVirtualName())
-            .withSize(deviceForType.getSize());
-
-    BlockDevice device2 =
-        new BlockDevice()
-            .withDeviceName(mapping.getDeviceName())
-            .withVirtualName(mapping.getVirtualName())
-            .withSize(mapping.getEbs().getVolumeSize());
-
-    return device1.equals(device2);
-  }
-
-  private Optional<ResolvedAmiResult> getAmiResult(
-      List priorOutputs, String region, String amiName) {
-    for (Object o : priorOutputs) {
-      if (o instanceof ResolvedAmiResult) {
-        ResolvedAmiResult result = ((ResolvedAmiResult) o);
-        if (result.getRegion().equals(region)
-            && (amiName.equals(result.getAmiId()) || amiName.equals(result.getAmiName()))) {
-          return Optional.of(result);
-        }
-      }
-    }
-
-    return Optional.empty();
+  @Nonnull
+  @Override
+  protected SagaFlow buildSagaFlow(List priorOutputs) {
+    this.priorOutputs = priorOutputs;
+    return new SagaFlow()
+        .then(FindAutoScalingGroup.class)
+        .then(PrepareModifyServerGroupLaunchTemplate.class)
+        .then(ModifyServerGroupLaunchTemplate.class)
+        .then(UpdateAutoScalingGroup.class);
```

Comment:
I think a dedicated saga exception handler is required or is the case that exceptions that are of type SpinnakerExceptions will be retried automatically?

Reply:
Yeah this is still in progress. Got a few todos regarding exception handling.